### PR TITLE
VACMS-12287 Add content model documentation entity.

### LINKED ISSUE #12287: Add content model documentation entity.
As a site maintainer I want a fieldable "documentation entity" that can entity reference a content type, paragraph, taxonomy, or field so that I am able to add fields to contain the information I need.

## AC's
- [ ] code committed to close [this Drupal issue](https://www.drupal.org/project/content_model_documentation/issues/3334959)

### PR DESCRIPTION
**Pull request by swirtSJW**
## Description

closes #12287

Code written in the upstream module for this issue https://git.drupalcode.org/project/content_model_documentation/-/compare/1.0.0-alpha3...1.0.0-alpha4?from_project_id=101712&straight=true

## QA steps

What needs to be checked to prove this works?  
What needs to be checked to prove it didn't break any related things?  
What variations of circumstances (users, actions, values) need to be checked?  
https://pr12432-i5lewscwe0rfmaekkrclnpbumnlficso.ci.cms.va.gov/
As an admin
- [ ] Go to https://pr12432-i5lewscwe0rfmaekkrclnpbumnlficso.ci.cms.va.gov/admin/config/system/cm_document validate that all entity types and fields are checked as documentable.
- [ ] validate that three other tabs show on for managing fields, display and form 
image
(these function, but will be tested on a separate issue so not testing here)
- [ ] Validate that there is a menu item for the settings page Configuration -> System -> Content Model Documentation
image
- [ ] Validate that there is a menu item for the Documentation list. Structure -> Content Model Documents  
- [ ]  Click the menu item Validate that it takes you here https://pr12432-i5lewscwe0rfmaekkrclnpbumnlficso.ci.cms.va.gov/admin/structure/cm_document
- The list will likely be empty unless someone else created documents.
- [ ] Click the "Add new Content Model Document" button.  validate that it takes you to https://pr12432-i5lewscwe0rfmaekkrclnpbumnlficso.ci.cms.va.gov/admin/structure/cm_document/add
- [ ] Pull down the "Documentation for" select list and validate that it contains fields for blocks, media, menus, nodes, paragraphs and taxonomy.
- [ ] Select one, add some text in Notes and press the save button, validate that it saves.  (you should see it appear in the list after save)
- [ ] Validate that you can click Name of the newly created document and be taken to the Document.
- [ ] Validate that the name of the Document matches the field you selected to document.
- [ ] Click the edit tab. Validate you can see the fields and be able to make an edit and save.
- Click on the name of the document in the list
- [ ] Validate that you see a tab for Revisions (the revisions list does not work yet it is a known bug)
- [ ] validate you see a clone tab and that it works
- [ ] validate that you see a delete tab and that it works
- [ ] Visit https://pr12432-i5lewscwe0rfmaekkrclnpbumnlficso.ci.cms.va.gov/admin/reports/dblog and validate that there are no scary messages related to your interactions.

   

### Definition of Done

- [x] Tests have been added if necessary (upstream).
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [x] `CMS Team`
- [ ] `Public websites`
- [ ] `Facilities`
- [ ] `User support`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [x] no.  Too soon to have developers using it yet.  Needs custom fields added.

### CODE REVIEW

**Comment by swirtSJW** [comment]
Proof of hook_update_n performance
image

**Comment by swirtSJW** [comment]
For reference, here is all the code that changed upstream as part of this issue
https://git.drupalcode.org/project/content_model_documentation/-/compare/1.0.0-alpha3...1.0.0-alpha4?from_project_id=101712&straight=true

**Review by EWashb — APPROVED** [review]
validated all steps. Worked as indicated. Nothing scary found

**Comment by laflannery** [comment]
On the Documentation list, there is a fieldset filter combo of an operator dropdown plus an input field. This input field does not have a label element associated with it:
image

However, I started looking in other places and this doesn't look specific to this task (It occurs on at least one other Facilities view as well) so I believe it would be out of scope here. Is that correct and should I open a separate backlog issue for this?

**Comment by swirtSJW** [comment]
@laflannery   I double checked the filter configuration to make sure there was no option that I neglected to set.  There was not, so yes this would be a Drupal Core Views issue that is outside the scope of this ticket.
image

My guess is there is likely an existing issue on drupal.org issue but it may take some digging to find it.  Here is the appropriate issue queue
https://www.drupal.org/project/issues/drupal?component=views.module

**Review by laflannery — APPROVED** [review]

**Comment by laflannery** [comment]
Thanks, I added a separate ticket (#12554) for the above.

File: docroot/modules/custom/va_gov_db/va_gov_db.install [hunk 1]
```diff
@@ -1104,3 +1105,44 @@ function va_gov_db_update_9012(&$sandbox) {
...
+ */
+function va_gov_db_update_9013(&$sandbox) {
+  $messages = '';
+  $schema = Database::getConnection()->schema();
+
+  // Because the content_model_documentation module now has a content type,
+  // the module needs to be uninstalled and then re-installed.  The problem is
+  // the cleanup tries to remove content type fields that do not exist and
+  // throws a fatal error.  So we have to create fake tables with the right
+  // fields so the cleanup on module uninstall can proceed.
```
Comment (ndouglas) [inline docroot/modules/custom/va_gov_db/va_gov_db.install:1120]: 😭 , but such is life.

**Comment by swirtSJW** [comment]
image

**Pull request merged**